File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    n = len(maps)
    m = len(maps[0])
    
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
    distance = [[-1 for _ in range(m)] for _ in range(n)]
    distance[0][0] = 1
    
    queue = deque([(0, 0)])
    
    while queue:
        x, y = queue.popleft()
        
        if x == n-1 and y == m-1:
            return distance[x][y]
        
        for dx, dy in directions:
            nx, ny = x + dx, y + dy
            
            if (0 <= nx < n and 0 <= ny < m and 
                maps[nx][ny] == 1 and distance[nx][ny] == -1):
                
                distance[nx][ny] = distance[x][y] + 1
                queue.append((nx, ny))
    
    return -1

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    public int solution(int[][] maps) {
        int n = maps.length;
        int m = maps[0].length;
        int[] mX = new int[] {1, 0, -1, 0};
        int[] mY = new int[] {0, 1, 0, -1};
        int[][] result = new int[n][m];
        for (int i = 0; i < n; i++) {
            Arrays.fill(result[i], Integer.MAX_VALUE);
        }
        result[0][0] = 1;
        
        Queue<Pos> q = new LinkedList<>();
        q.offer(new Pos(0, 0));
        
        while(!q.isEmpty()) {
            Pos cur = q.poll();
            
            for (int i = 0; i < 4; i++) {
                int nX = cur.x + mX[i];
                int nY = cur.y + mY[i];
                
                if (nX < 0 || nY < 0 || nX >= n || nY >= m) continue;
                if (maps[nX][nY] == 0) continue;
                if (result[nX][nY] > result[cur.x][cur.y] + 1) {
                    result[nX][nY] = result[cur.x][cur.y] + 1;
                    q.offer(new Pos(nX, nY));
                }
            }
        }
        
        return result[n - 1][m - 1] == Integer.MAX_VALUE ? -1 : result[n - 1][m - 1];
    }
    class Pos {
        int x;
        int y;
        
        Pos(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }
}